### DISCUSSION TITLE
Trained-vehicle model implementation

### DISCUSSION THREAD

**Discussion by ibrahimbiyikli** [eleurent/highway-env#222] [Ideas]
First of all, thank you for creating highway-env. I obtained a trained DQN agent model in a highway environment. I want to train a new  DQN agent in pre-trained  DQN agent models. How can we put trained models in the highway environment? Briefly, I want to exchange IDM vehicles with trained agent models for training new agents.

**Comment by eleurent** [comment]
Hi @ibrahimbiyikli,

That's a good idea, but it requires a bit of development. Here are the main steps of how it can be done:
1. Create a new TrainedVehicle class, which inherits from `MDPVehicle`, for instance.
2. In the __init__, load a trained model from a path
3. Overload the `act()` method to:
  3.1 Generate an observation (in the same format as that which was used to train the model), centered on the vehicle (self)
  3.2 Feed the observation to the model, get the resulting action, and execute it by forwarding it to the `MDPVehicle` class through `super().act(action)`

**Reply by ibrahimbiyikli** [reply]
Hi @eleurent ,

Firstly thank you for your response, 

I have some questions. I tried to load the model but I could not understand how model loading works. Also, how should I use the model to feed it with observation? If you explain, I would be happy.

**Reply by eleurent** [reply]
Hi @ibrahimbiyikli 
The answer to these questions (how to load a model, how to feed a model with an observation) depend on which RL framework you are using. Generally speaking, it should probably something looks something like

```
from your_rl_library import YourDQNModel


model = YourDQNModel()
model.load(path_to_saved_model)

# And then something like
for vehicle in env.road.vehicles:
    # generate the observatoin
    obs = env.observation.observe(vehicle)
    # feed the obs through the model to get an action
    action = model.predict(obs)
    # make the vehicle do the action
    vehicle.act(action)
```

but that is just some gross pseudo-code, you will have to adapt it to your particular use-case.

**Reply by eleurent** [reply]
But your questions (loading a model, feeding an observation to a model) should probably be transferred to the DeepRL library that you use, they are unrelated to highway-env or any specific environment.
